evennia/server/portal/telnet.py

```diff
@@ -440,7 +440,8 @@ def send_text(self, *args, **kwargs):
                     prompt = mxp_parse(prompt)
             prompt = to_bytes(prompt, self)
             prompt = prompt.replace(IAC, IAC + IAC).replace(b"\n", b"\r\n")
-            prompt += IAC + GA
+            if not self.protocol_flags.get("NOGOAHEAD", True) and not self.protocol_flags.get("NOPROMPTGOAHEAD", True):
+                    prompt += IAC + GA
```

Comment:
Hm. If `NOGOAHEAD=True` we *don't* want to send a GA for prompts though, right? The client accepted that we wont send a GA so makes sending one superfluous.

`NOGOAHEAD=True` with `NOPROMPTGOAHEAD=True` is how it is now.
`NOGOAHEAD=False` (negotiated with client) with `NOPROMPTGOAHEAD=False` (set by ttype = MUDLET) is what we're targetting here.
`NOGOAHEAD=True` with `NOPROMPTGOAHEAD=False` would not be settable outside of manually fiddling with options, buy given we agreed to not send GA's, sending them seems a bit rude.

It may make more sense to just have three values for NOGOAHEAD... True and False as negotiated by the client, but then a value of "PROMPT" for 'send GA on prompt only'. Mixes bool and strings though which is a little dirty.

Reply:
All that being said, your example works for all of the above anyway...

`True`, `True` does not send GA ✅
`False`, `False` sends GA on prompt ✅
`False`, `True` sends GA on prompt ✅
`True`, `False` does not send GA ✅

Reply:
`WILL/WONT GA` is a standardized telnet protocol handshake that we can't change with a third value. There is no information about prompts or anything there. So the `NOPROMPTGOAHEAD` must by necessity be something we set manually (or hard-code for mudlet, in this case). 

Before you wrote this: 
> Mudlet contains a isPrompt() function to detect prompts. This is powered by the presence of a GA on the line that sends it 
> [...] Without a GA on every line Mudlet (latest version) doesn't render anything. It needs the end line in its absence, but also treats any line with GA as a prompt.

From this I conclude that mudlet wants `NOGOAHEAD` for regular text, but `GA` for prompts (using `FORCEDENDLINE` to identify the end of regular text instead). At least that's how I understood it?

Reply:
Exactly. So NOGOAHEAD=False, NOPROMPTGOAHEAD=False and FORCEDENDLINE=True.

Reply:
No. `NOGOAHEAD = True`, `NOPROMPTGOAHEAD=False` and `FORCEDENDLINE=True` based on what you said. If `NOGOAHEAD/NOPROMPTGOAHEAD` were always the same, why would we even need the latter?

Reply:
Ok. Starting from the top and in an effort to simplify...

If `NOGOAHEAD=True`, that means we suppress go aheads always. We do not send a GA under any circumstances.

If `NOGOAHEAD=False` we defer to NOPROMPTGOAHEAD to decide when to send one, thus:

* If `NOPROMPTGOAHEAD=True` we always send a GA (This what Mudlet does not like)

* If `NOPROMPTGOAHEAD=False` we only send a GA on a prompt (sendLine does not terminate in a GA)

I think we're actually saying the same thing but clashing on what NOPROMPTGOAHEAD means (for me its do we send a GA on something that is not a prompt).

Reply:
I think the disparity is rather that I'm seeing it like this:

- `NOGOAHEAD=True/False` controls the GA state for the regular text output.
- `NOPROMPTGOAHEAD` controls the GA state for prompts only. This means that this can be `False/True` and still send `GA` even if `NOGOAHEAD=True`. 

I guess both approaches work, but it seems strange that `NOPROMPTGOAHEAD` would control non-prompt output when its very name indicates it's prompt-related.

Reply:
I do see your point but that would mean that to suppress GA on a regular (not a prompt) output, we'd need to change the negotiated NOGOAHEAD value to True for Mudlet, and set NOPROMPTGOAHEAD to False. Is that what we want?

Reply:
I would imagine mudlet could send this itself, but yes, this sounds like the most logical solution, since NOGOAHEAD=True is really the most commonly expected setting.

Reply:
I'm sure it could but by default it wont, and expects the GA-lite experience (only GA's on prompts rather than after everything). Latest push makes it work as described above.